example/poisson_gmg/poisson_package.hpp

```diff
@@ -0,0 +1,406 @@
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef EXAMPLE_POISSON_GMG_POISSON_PACKAGE_HPP_
+#define EXAMPLE_POISSON_GMG_POISSON_PACKAGE_HPP_
+
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <kokkos_abstraction.hpp>
+#include <parthenon/package.hpp>
+
+#define VARIABLE(ns, varname)                                                            \
+  struct varname : public parthenon::variable_names::base_t<false> {                     \
+    template <class... Ts>                                                               \
+    KOKKOS_INLINE_FUNCTION varname(Ts &&...args)                                         \
+        : parthenon::variable_names::base_t<false>(std::forward<Ts>(args)...) {}         \
+    static std::string name() { return #ns "." #varname; }                               \
+  }
+
+namespace poisson_package {
+using namespace parthenon::package::prelude;
+// GMG fields
+VARIABLE(poisson, res_err);
+VARIABLE(poisson, D);
+VARIABLE(poisson, u);
+VARIABLE(poisson, rhs);
+VARIABLE(poisson, u0);
+VARIABLE(poisson, temp);
+VARIABLE(poisson, exact);
+VARIABLE(poisson, rhs_base);
+
+// BiCGStab fields
+VARIABLE(poisson, rhat0);
+VARIABLE(poisson, v);
+VARIABLE(poisson, h);
+VARIABLE(poisson, s);
+VARIABLE(poisson, t);
+VARIABLE(poisson, x);
+VARIABLE(poisson, r);
+VARIABLE(poisson, p);
+
+constexpr parthenon::TopologicalElement te = parthenon::TopologicalElement::CC;
+
+std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin);
+
+template <class in, class out, bool only_md_level = false>
+TaskStatus CopyData(std::shared_ptr<MeshData<Real>> &md) {
+  using TE = parthenon::TopologicalElement;
+  auto pmb = md->GetBlockData(0)->GetBlockPointer();
+  IndexRange ib = pmb->cellbounds.GetBoundsI(IndexDomain::entire, te);
+  IndexRange jb = pmb->cellbounds.GetBoundsJ(IndexDomain::entire, te);
+  IndexRange kb = pmb->cellbounds.GetBoundsK(IndexDomain::entire, te);
+
+  int nblocks = md->NumBlocks();
+  std::vector<bool> include_block(nblocks, true);
+  if (only_md_level) {
+    for (int b = 0; b < nblocks; ++b)
+      include_block[b] =
+          (md->grid.logical_level == md->GetBlockData(b)->GetBlockPointer()->loc.level());
+  }
+
+  auto desc = parthenon::MakePackDescriptor<in, out>(md.get());
+  auto pack = desc.GetPack(md.get(), include_block);
+  parthenon::par_for(
+      DEFAULT_LOOP_PATTERN, "SetPotentialToZero", DevExecSpace(), 0,
+      pack.GetNBlocks() - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+      KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
+        pack(b, te, out(), k, j, i) = pack(b, te, in(), k, j, i);
```

Comment:
I don't understand the comment about doing this by metadata?

I think I tried this variadically with fold expressions (or at least in a very similar kernel). It worked fine on CPU, but the GPU compiler complained.

Reply:
> I don't understand the comment about doing this by metadata?

You could pull out metadata for, e.g., (poisson) variables, rather than pulling out variables of a type. Something like that. Would require adding custom metadata for that purpose, though.

> I think I tried this variadically with fold expressions (or at least in a very similar kernel). It worked fine on CPU, but the GPU compiler complained.

:+1: